[perf] validation perf work

jshaughn opened the pull request:
Part-of #8007 

So this PR does not fundamentally change the way we do validations, and therefore is looking to make only marginal perf gains via some code refactoring for efficiency.  Perhaps there are fundamental things we can do to speed things up, but I'm not yet sure what they may be.  The biggest win may be to just invest in reconciling in response to config changes, as opposed to looking for gains in approach.  One way to do that would be to have watchers and respond to changes. But perhaps an easier way would be to "hash" all of the resource versions of everything involved in validation, save that, and skip validation if it doesn't change.  We'd still have to fetch everything on each reconcile, but most everything is cached.

As for this PR, the general approach to the refactor was to "pull up" anything that could be calculated or fetched one time, and then re-used.  The existing code buried a fair amount of calculation in functions that were called many times in one reconciliation, as there is a *lot* of looping going on.  Despite caching, the repeated lookup and calculation can add up in larger deployments.

This introduces an `validationInfo` object, `vInfo`, that accumulates information used for the validation pass.  It gets allocated up front via a `NewValidationInfo` call, and has three parts: base, cluster, namespace.  Base information can be calculated one time is stored in the initial `New`.  Cluster information must be reset for each cluster being processed, and Namespace info for each Namespace.  The `vInfo` object is then passed around throughout the process, and used to define the checkers.  Note, to minimize changes I stopped short of passing it into the checkers themselves.

I did some testing with `business/istio_validations_perf_test.go` setting all the vars to 350, and it showed about a 16% gain. That's not Earth-shattering, and I'm not sure how much it reflects reality, but it would be a decent improvement.

jshaughn commented:
> Did a run on Jenkins env, there is up to 40% perf improvement, and all pages are improved:

OOh, thanks @hhovsepy , great data!  I guess I'll open this up for review!

hhovsepy commented:
Tested locally, LGTM, I like the solutions of grouping the attributes and skipping creating variables in loops repetitively.

nrfox commented on business/istio_validations.go:
```diff
@@ -414,155 +522,100 @@ func runObjectReferenceChecker(referenceChecker ReferenceChecker) models.IstioRe
-					}
-					if !found {
-						serviceAccounts[cluster] = append(serviceAccounts[cluster], saFullName)
+	for _, ns := range namespaces {
+		workloadList := workloadsMap[ns.Name]
+		for _, wl := range workloadList.Workloads {
+			for _, sAccountName := range wl.ServiceAccountNames {
+				saFullName := fmt.Sprintf("%s/ns/%s/sa/%s", istioDomain, ns.Name, sAccountName)
+				found := false
+				for _, name := range serviceAccounts {
+					if name == saFullName {
+						found = true
+						break
 					}
 				}
```
Do we need the outermost loop over namespaces? Do we need namespaces at all if all we want to do is return a list of the service accounts for all the workloads?

jshaughn replied:
We are generating the SA full name,  and looking at it I can see that we use the namespace name in the `saFullName` string generation.  But you are right, if we can assume that the workspace's namespace is the same as what we get from the outer loop, then I guess it's not necessary.  Having said that, I don't think a change computationally saves us anything, as in the end we're still processing each workload one time, right?  There is one trivial saving we could make here. If we keep it as is we can actually construct the `saFullName` prefix one time per namespace, Like this:

```
	for _, ns := range namespaces {
	        saFullNameNs := fmt.Sprintf("%s/ns/%s/sa/", istioDomain, ns.Name)
		workloadList := workloadsMap[ns.Name]
		for _, wl := range workloadList.Workloads {
			for _, sAccountName := range wl.ServiceAccountNames {
				saFullName := fmt.Sprintf("%s%s", saFullNameNs, sAccountName)
				found := false
                                etc...
```

Do you agree?  I can make that change...

nrfox replied:
> if we can assume that the workspace's namespace is the same as what we get from the outer loop

We can always assume this. Workloads are bound to a single namespace. You should be able to get the namespace on the workload object itself.

> Having said that, I don't think a change computationally saves us anything, as in the end we're still processing each workload one time, right?

The total times we are looping is `namespaces * workloadList.Workloads * wl.ServiceAcccountNames` so if we can eliminate `namespaces` then it's just `workloadList.Workloads * wl.ServiceAccountNames` which I think will save us something somewhat significant esp. with many namespaces.

Is all we are doing here constructing the "saFullName"? Maybe that's something that should live in the Workload object itself and gets set when you Parse the workload?

jshaughn replied:
Let me look, that may be a nice optimization, as it would allow us to completely remove this from `vInfo`...

jshaughn replied:
As for, `The total times we are looping is namespaces * workloadList.Workloads * wl.ServiceAcccountNames`, I'm not sure I agree, because we don't repeat all workloads for each namespace.  Either way, it could be moot if we do this other suggestion.

jshaughn replied:
Looking at this, I think we still need the vInfo.saMap, because it is passed directly into the checkers, and apparently does require cross-cluster.  I expressly don't want to touch the checkers themselves, at least not in this PR, only later if we can prove that the checking logic must also be improved.  But, we don't necessarily have to reconstruct all of the strings on each reconcile if we instead store it on the workload at parse-time. Let me go down that path a little and see if it has any gotchas.  I'm not sure how much a benefit this will be, but I'll proceed if it's easy enough...

jshaughn replied:
OK, beating this to death...  I think it's too disruptive to store the sa's in two formats on the workload, just for this one use case and a very nominal compute-gain.  So, I'm just going to leave this in here.  But, I will optimize it a bit and looking more carefully at it, I think maybe there is a bug here as well.

jshaughn replied:
I pushed the change, will let CI verify that everything still looks good.  By the way, Kudos to @hhovsepy or whomever wrote some of these tests. One failure drove me to the brink of insanity, but in the end it was a real issue.

jshaughn replied:
well you spotted the same bug.  See the updated code, I also moved to a map :-)   I'm leaving the outer namespace loop because it allows for a minor improvement to generate a prefix that can be applied below.

jshaughn replied:
If you really can't stand the looping feel free to change it in a future PR :-)  I want to put this one in the can and move on to hopefully bigger fish...

nrfox replied:
*insert infinite loop gif

No this is fine. And all the changes together are great improvement.

nrfox commented:
> Did a run on Jenkins env, there is up to 40% perf improvement, and all pages are improved: ![cbbeb567-af05-48f2-b7b2-3161c195a948](https://private-user-images.githubusercontent.com/604313/418019356-442c3471-deb3-4bbf-ba23-0c6e0595315c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwMTE1NzUsIm5iZiI6MTc0MTAxMTI3NSwicGF0aCI6Ii82MDQzMTMvNDE4MDE5MzU2LTQ0MmMzNDcxLWRlYjMtNGJiZi1iYTIzLTBjNmUwNTk1MzE1Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMwM1QxNDE0MzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMGJkYjM3YTEyZDYxZWI4MGU5OGZmNjE1MWRkMDFmZGJjZGRjYjBkMjRjOWMyZjFjYTJjNWQ2YzhkYzY5YTdmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OwmqMfskPghn5Wq3GrM_5CXNgi7l0gd3Xs_0X1DKj-8)

Glad we are addressing some of these issues. It really takes 10s to load the graph?

hhovsepy commented:
> 
> Glad we are addressing some of these issues. It really takes 10s to load the graph?

This is the perf env, with 50 namespaces, selecting all of them and checking idle nodes and edges.

jshaughn commented:
> Glad we are addressing some of these issues. It really takes 10s to load the graph?

> This is the perf env, with 50 namespaces, selecting all of them and checking idle nodes and edges.

The graph generation is fundamentally designed to generate a graph per requested namespace, and then stitch it together.  So, as the number of namespaces goes up, the more sub-graphs get generated, the more queries to Prom, the more appender runs, etc...  I've looked several times at changing it to grab everything from Prom (basically everything for the time period, for all namespaces, in one fell swoop, and then deal with it all in the generator.  While that would reduce Prom queries, I think it has a variety of drawbacks around complexity, and possibly perf.  I'm not sure it would result in a better experience and the amount of work is daunting.

jshaughn commented on business/istio_validations.go:
```diff
@@ -119,10 +119,81 @@ func (in *IstioValidationsService) GetValidationsForWorkload(ctx context.Context
+	rbacDetails *kubernetes.RBACDetails //
+}
+
+type validationClusterInfo struct {
+	cluster          string                        // the cluster being validated
+	istioConfig      *models.IstioConfigList       // config for the cluster (all namespaces)
+	registryServices []*kubernetes.RegistryService // registry services for the cluster (all namespaces)
+}
```
Because this info is replaced on each cluster iteration, and analogously, each namespace iteration, I **think** it makes sense to have the separate structures?  The idea is that `validationInfo` holds the "base" info that is usable across the reconciliation pass, and then `validationClusterInfo` holds what can be used across the cluster's reconciliation, and finally, `validationNamespaceInfo`, across the namespace's reconciliation.

What I am pondering, in a possible "part two" of this optimization work, is that prior to actually running the checkers, we check against the prior run to see if there is a change in the istio objects'  resourceVersions (basically a hash).  So, we'd gather the info required to run the checkers, but only do it if we detect a change.

nrfox replied:
> What I am pondering, in a possible "part two" of this optimization work, is that prior to actually running the checkers, we check against the prior run to see if there is a change in the istio objects' resourceVersions (basically a hash). So, we'd gather the info required to run the checkers, but only do it if we detect a change.

Yes we can do that. You setup the reconcile loop to watch for any changes to some subset of objects.

nrfox commented on business/istio_validations.go:
```diff
@@ -119,10 +119,81 @@ func (in *IstioValidationsService) GetValidationsForWorkload(ctx context.Context
+// validationInfo holds information gathered during a single validation reconciliation. It is used to hold information that
+// may otherwise need to be recalculated.
+type validationInfo struct {
+	// cross-cluster information
+	clusters []string                                  // all clusters being validated
+	mesh     *models.Mesh                              // control plane info
+	nsMap    map[string][]models.Namespace             // cluster => namespaces
+	saMap    map[string][]string                       // cluster => serviceAccounts
+	wlMap    map[string]map[string]models.WorkloadList // cluster => namespace => WorkloadList, all workloads
```
Would look like this where you would create a `vInfo` for that cluster.
```
	vInfo, err := r.validationsService.NewValidationInfo(ctx, cluster)
	if err != nil {
		log.Errorf("[ValidationsReconciler] Error creating validation info: %s", err)
		return ctrl.Result{}, err
	}
	
       clusterValidations, err := r.validationsService.Validate(ctx, cluster, vInfo)
```

I'm assuming here that we are not doing cross cluster validations. But looking at this comment it looks like we're already doing cross cluster validation of service account info?
```
	// validation requires cross-cluster service account information.
```

nrfox replied:
Yeah this is just preference and it looks like there's already the need for cross cluster validations so keep it as is.
